internal/utils/time.go

```diff
@@ -59,3 +59,53 @@ func RunOnInterval(ctx context.Context, fn func(), interval time.Duration) {
 		}
 	}()
 }
+
+// RepeatPolicy repeat behavior for RunOnIntervals Function
+type RepeatPolicy int
+
+const (
+	// RepeatAll repeats all intervals
+	RepeatAll RepeatPolicy = iota
+	// RepeatLast repeats only the last interval
+	RepeatLast
+	// RepeatNone does not repeat
+	RepeatNone
+)
+
+// RunOnIntervals the given function once on the specified intervals, and
+// repeat according to the supplied RepeatPolicy.
+func RunOnIntervals(ctx context.Context, fn func(), intervals []time.Duration, repeatPolicy RepeatPolicy) {
+	if len(intervals) < 1 {
+		return
+	}
+	// copy intervals
+	intv := intervals[:]
+
+	// initialize timer
+	timer := time.NewTimer(intv[0])
+	go func() {
+		defer timer.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+
+				if len(intv) > 1 {
+					// advance the interval
+					intv = intv[1:]
+				} else {
+					// evaluate repeat policies
+					if repeatPolicy == RepeatNone {
+						return
+					} else if repeatPolicy == RepeatAll {
+						intv = intervals[:] // copy the original interval list
```

Comment:
Don't assume the original intervals slice's elements are immutable.  A slice only points to a backing array and that can change due to logic outside of this function that reuses the `intervals` slice that was passed in, such as reassigning distinct elements in the slice.  You may not be doing that in the monitor code now, but it is a potential source of a very tricky bug.

I would just stick with the copy in `intv` that you have and use an index variable to keep track of where you are in it instead of reassigning it.  That copy is good because it happens before the function returns.  If it is `RepeatLast` you can just leave the index stuck at the end forever.